cpp/serve/request_state.cc

```diff
@@ -116,13 +118,14 @@ RequestStateEntry::RequestStateEntry(
 
 DeltaRequestReturn RequestStateEntryNode::GetReturnTokenIds(const Tokenizer& tokenizer,
                                                             int max_single_sequence_length) {
+  // Disable this case because speculative may have remaining draft tokens.
   // - Case 0. There is remaining draft output ==> Unfinished
   //   All draft outputs are supposed to be processed before finish.
-  for (RequestModelState mstate : this->mstates) {
-    if (!mstate->draft_output_tokens.empty()) {
-      return {{}, {}, Optional<String>()};
-    }
-  }
+  // for (RequestModelState mstate : this->mstates) {
+  //   if (!mstate->draft_output_tokens.empty()) {
+  //     return {{}, {}, Optional<String>()};
+  //   }
+  // }
```

Comment:
Does this mean that, the batch verification stage may not clear all the draft tokens, so that there are new committed tokens and also remaining draft tokens?

Reply:
Yes. Eagle requires us to use the previous token's hidden states to predict the next token. We record the hidden states via AddDraftToken so that in batch draft step we can get the hidden states. This requries the batch verify step to prepare the next draft token. As a result,`draft_output_tokens` is not empty.